server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -1638,6 +1638,15 @@ private SearchSourceBuilder parseXContent(
         }
 
         knnSearch = knnBuilders.stream().map(knnBuilder -> knnBuilder.build(size())).collect(Collectors.toList());
+        if (rankBuilder != null) {
```

Comment:
So, I have a stupid question, do we validate that you cannot provide other things when retrievers are provided? Meaning, its invalid to provide `rank` and a `retriever`? Or `query` and a `retriever`?

Reply:
This takes place at a later time through `SearchSourceBuilder#validate`. So, in practice we would now validate the search request as if a retriever was provided from the get go. 

But if there was an initial misconfiguration with providing both a `retriever` and a `rank` then we would just override the retriever at this point, whereas it would probably be best to throw (will update to account for this).